cleverhans/dataset.py

```diff
@@ -0,0 +1,144 @@
+"""Dataset class for CleverHans
+
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from cleverhans.utils_mnist import data_mnist
+
+keras = None  # Only load keras if user tries to use a dataset that requires it
+
+
+class Dataset(object):
+    """Abstract base class representing a dataset.
+    """
+
+    # The number of classes in the dataset. Should be specified by subclasses.
+    NB_CLASSES = None
+
+    def __init__(self, **kwargs):
+        if "self" in kwargs:
+            del kwargs[self]
+        self.kwargs = kwargs
+
+    def get_factory(self):
+        """Returns a picklable callable that recreates the dataset.
+        """
+
+        if hasattr(self, 'kwargs'):
+            kwargs = self.kwargs
+        else:
+            kwargs = {}
+
+        return Factory(type(self), kwargs)
+
+
+class MNIST(Dataset):
+    """The MNIST dataset"""
+
+    NB_CLASSES = 10
+
+    def __init__(self, train_start=0, train_end=60000, test_start=0,
+                 test_end=10000, center=False):
+        super(MNIST, self).__init__(**locals())
+        self.kwargs = locals()
+        x_train, y_train, x_test, y_test = data_mnist(train_start=train_start,
+                                                      train_end=train_end,
+                                                      test_start=test_start,
+                                                      test_end=test_end)
+
+        if center:
+            x_train = x_train * 2. - 1.
+            x_test = x_test * 2. - 1.
+
+        self.x_train = x_train
+        self.y_train = y_train
+        self.x_test = x_test
+        self.y_test = y_test
+
+
+class CIFAR10(Dataset):
+    """The CIFAR-10 dataset"""
+
+    NB_CLASSES = 10
+
+    def __init__(self, train_start=0, train_end=60000, test_start=0,
+                 test_end=10000, center=False):
+        super(CIFAR10, self).__init__(**locals())
+        packed = data_cifar10(train_start=train_start,
+                              train_end=train_end,
+                              test_start=test_start,
+                              test_end=test_end)
+        x_train, y_train, x_test, y_test = packed
+
+        if center:
+            x_train = x_train * 2. - 1.
+            x_test = x_test * 2. - 1.
+
+        self.x_train = x_train
+        self.y_train = y_train
+        self.x_test = x_test
+        self.y_test = y_test
+
+
+class Factory(object):
+    """
+    A callable that creates an object of the specified type and configuration.
+    """
+
+    def __init__(self, cls, args, kwargs):
```

Comment:
Note, you can remove args here since it's unused.

Reply:
thanks, done